### PR TITLE
pkg/ipam: fix multi-pool allocator not releasing un-used /32 and /128 CIDRs

### LINKED ISSUE #40350: Multi-Pool IPAM sometimes does not release pool pod CIDRs after Pod is
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.17.5 and lower than v1.18.0

### What happened?

We are observing an issue that is similar to https://github.com/cilium/cilium/issues/37746 which was fixed in v1.17.2 but are still seeing the issue in v1.17.5. Pods are running fine at first but _sometimes_ when they are removed/recreated and end up on a different host, they will not be able to get an IP and stay in the ContainerCreating forever. This affects some scaling and upgrades.

I am expecting that when the pod is redeployed on another host that the IPs used by the pod are re-used by the pod on the new host.

The pod status shows
```
 Warning  FailedCreatePodSandBox  2m51s (x351 over 80m)  kubelet  (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "3872fecd03c3b878a5a89002a899e3071b25884f77e5f0345d9f574ed01a70f8": plugin type="multus-shim" name="multus-cni-network" failed (add): CmdAdd (shim): CNI request failed with status 400: 'ContainerID:"3872fecd03c3b878a5a89002a899e3071b25884f77e5f0345d9f574ed01a70f8" Netns:"/var/run/netns/cni-ee8951e0-8410-211d-0080-d3ebe9df7e72" IfName:"eth0" Args:"K8S_POD_NAMESPACE=my-ns;K8S_POD_NAME=my-pod-6d5b7cdb6-2nmk4;K8S_POD_INFRA_CONTAINER_ID=3872fecd03c3b878a5a89002a899e3071b25884f77e5f0345d9f574ed01a70f8;K8S_POD_UID=24a1bf8a-dc9a-4a0c-938c-ddf3bd3708cb;IgnoreUnknown=1" Path:"" ERRORED: error configuring pod [my-ns/my-pod-6d5b7cdb6-2nmk4] networking: [my-ns/my-pod-6d5b7cdb6-2nmk4/24a1bf8a-dc9a-4a0c-938c-ddf3bd3708cb:cilium]: error adding container to network "cilium": plugin type="cilium-cni" failed (add): unable to allocate IP via local cilium agent: [POST /ipam][502] postIpamFailure "unable to allocate from pool \"my-pool\" (family ipv6): pool not (yet) available"
': StdinData: {"clusterNetwork":"/host/etc/cni/net.d/05-cilium.conflist","cniVersion":"0.3.1","logLevel":"verbose","logToStderr":true,"name":"multus-cni-network","type":"multus-shim"}
```

The operator shows the following message. It can indicate ipv4 or ipv6; depends on what is stuck.
```
level=warning msg="Failed to allocate PodCIDRs to node" error="failed to allocate ipv6 address for node \"hostname\" from pool \"my-pool\": pool empty" nodeName=hostname subsys=ipam-allocator-multi-pool
```

`kubectl get ciliumnode` shows 2 ips assigned to the old host the pod was running on when it should only have 1 of each ipv4 and ipv6.
.spec.ipam.pools.allocated:
```
    {
      "cidrs": [
        "1.2.3.4/32",
        "1.2.4.6/32",
        "2001:1234:5678:9012::1/128",
        "2001:1234:5678:9012::2/128"
      ],
      "pool": "my-pool"
    },
```
.spec.ipam.pools.requested:
```
    {
      "needed": {
        "ipv4-addrs": 1,
        "ipv6-addrs": 1
      },
      "pool": "my-pool"
    },

```

### How can we reproduce the issue?

I have a CiliumPodIPPool setup with a couple of v4/v6 CIDRs.
```
---
apiVersion: cilium.io/v2alpha1
kind: CiliumPodIPPool
metadata:
  name: my-pool
spec:
  ipv4:
    cidrs:
      - 1.2.3.4/32
      - 1.2.4.6/32
    maskSize: 32
  ipv6:
    cidrs:
      - "2001:1234:5678:9012::1/128"
      - "2001:1234:5678:9012::2/128"
    maskSize: 128
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-deploy
  namespace: my-ns
  labels:
    app: my-app
spec:
  replicas: 2
  selector:
    matchLabels:
      app: my-app
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app: my-app
      annotations:
        ipam.cilium.io/ip-pool: my-pool
    spec:
      containers:
        - name: my-pod
          image: mycr.com/app:0.0.7

```
Kill one of the pods a few times and it will eventually get into the bad state. Last time I had the kill the pods around 5 times with around 30s between each kill to reproduce the error. It does seem to happen with longer times between the redeploys.

### Cilium Version

We've noticed this problem on various versions of Cilium 1.17.x but we checked on the latest v1.17.5 to make sure.
```
cilium-cli: v0.18.4 compiled with go1.24.3 on linux/amd64
cilium image (default): v1.17.4
cilium image (stable): v1.17.5
cilium image (running): 1.17.5
```

### Kernel Version

We build our own kernel 
`Linux <nodename> 6.12.0-<company>-366166 #366166~bpo11+1 SMP PREEMPT_DYNAMIC Mon Jan 27 20:45:15 UTC 2025 x86_64 GNU/Linux`

### Kubernetes Version

v1.31.4+rke2r1

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

```shell

```

### Anything else?

I am able to work around this issue by patching the CiliumNode and removing the unused IPs from `spec.ipam.pools.allocated[n]cidrs` list.

helm values
```
autoDirectNodeRoutes: false
bandwidthManager:
  bbr: true
  enabled: true
bgpControlPlane:
  enabled: true
bpf:
  datapathMode: netkit
  events:
    drop:
      enabled: true
    policyVerdict:
      enabled: true
    trace:
      enabled: false
  lbExternalClusterIP: true
  lbModeAnnotation: false
  masquerade: true
chainingMode: portmap
cluster:
  id: 34
  name: staging
cni:
  exclusive: false
devices: dummy0,dummy1,ethA,ethB
enableIPv4Masquerade: true
enableIPv6Masquerade: true
encryption:
  enabled: true
  type: wireguard
endpointRoutes:
  enabled: true
envoy:
  enabled: false
externalIPs:
  enabled: true
hostPort:
  enabled: true
hubble:
  enabled: true
  listenAddress: :4244
  metrics:
    dashboards:
      enabled: false
      namespace: cilium
    serviceMonitor:
      enabled: false
  preferIpv6: true
  relay:
    enabled: true
    prometheus:
      serviceMonitor:
        enabled: false
  ui:
    enabled: true
ipMasqAgent:
  enabled: false
ipam:
  mode: multi-pool
  operator:
    autoCreateCiliumPodIPPools:
      default:
        ipv4:
          cidrs:
            - 172.31.0.0/18
          maskSize: 24
        ipv6:
          cidrs:
            - 2001:1234:abcd:5678::/56
          maskSize: 64
ipv4:
  enabled: true
ipv4NativeRoutingCIDR: 0.0.0.0/0
ipv6:
  enabled: true
ipv6NativeRoutingCIDR: ::/0
k8sServiceHost: rke2-server.staging.tld.com
k8sServicePort: 6443
kubeProxyReplacement: true
l2NeighDiscovery:
  enabled: false
loadBalancer:
  algorithm: maglev
  dsrDispatch: geneve
  mode: hybrid
localRedirectPolicy: false
maglev:
  hashSeed: OsC/VXl9Yrtu3f2z
  tableSize: 509
nodePort:
  enableHealthCheck: false
  enabled: true
routingMode: native
socketLB:
  enabled: true
tunnelProtocol: geneve
```

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by alimehrabikoshki**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

releaseExcessCIDRsMultiPool interprets its argument as how many free addresses must remain after a CIDR is dropped, whereas neededIPs.IPvXAddrs is the total number of addresses required to be allocated.

See the condition `totalFree-free >= neededIPs` in releaseExcessCIDRsMultiPool.

In practice this means that if any other pod is still running, the CIDR that should be release has the stats: 
totalFree = 1, free = 1, neededIPs = 1 -> CIDR will be retained as `1 - 1 >= 1` == `false`

Conversely, after the last pod IP from the pool is released: totalFree = 2, free = 1, neededIPs = 0 -> CIDR can be dropped

By removing inUse IPs from neededIPs, empty CIDRs can be released immediately.

Below log lines (custom log lines) show the relevant counters when one out of the two /32's on the node is released but the CIDR is retained (.3/32 is still in use, .4/32 is not). 

```
source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:200 msg="Checking pod CIDR" module=agent.controlplane.ipam cidr=10.9.237.3/32 free=0 used=1 totalFree=1 neededIPs=1 free-after-remove=1
source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:200 msg="Checking pod CIDR" module=agent.controlplane.ipam cidr=10.9.237.4/32 free=1 used=0 totalFree=1 neededIPs=1 free-after-remove=0
source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:213 msg="Pod CIDR retained" module=agent.controlplane.ipam count=2 allocators="[0xc000e87710 0xc000e87770]"
```

After the last /32 is released:
```
time=2025-07-07T15:04:03.737839047Z level=debug source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:200 msg="Checking pod CIDR" module=agent.controlplane.ipam cidr=10.9.237.4/32 free=1 used=0 totalFree=2 neededIPs=0 free-after-remove=1
time=2025-07-07T15:04:03.73835463Z level=debug source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:200 msg="Checking pod CIDR" module=agent.controlplane.ipam cidr=10.9.237.3/32 free=1 used=0 totalFree=1 neededIPs=0 free-after-remove=0
source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:213 msg="Pod CIDR retained" module=agent.controlplane.ipam count=0 allocators=[]
```

Fixes: #40350 

```release-note
pkg/ipam: fix multi-pool allocator not releasing un-used /32 and /128 CIDRs
```

### CODE REVIEW

**Comment by alimehrabikoshki** [comment]
with this change:

```
source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:200 msg="Checking pod CIDR" module=agent.controlplane.ipam cidr=10.9.237.7/32 free=0 used=1 totalFree=1 neededIPs=0 free-after-remove=1
source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:200 msg="Checking pod CIDR" module=agent.controlplane.ipam cidr=10.9.237.3/32 free=0 used=1 totalFree=1 neededIPs=0 free-after-remove=1
source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:200 msg="Checking pod CIDR" module=agent.controlplane.ipam cidr=10.9.237.1/32 free=1 used=0 totalFree=1 neededIPs=0 free-after-remove=0
source=/go/src/github.com/cilium/cilium/pkg/ipam/pool.go:213 msg="Pod CIDR retained" module=agent.controlplane.ipam count=2 allocators="[0xc000921ec0 0xc001195f50]"
```

The unused /32 is removed even though there are two other CIDRs/allocators

File: pkg/ipam/multipool.go [hunk 1]
```diff
@@ -552,7 +552,14 @@ func (m *multiPoolManager) updateLocalNode(ctx context.Context) error {
...
 		if v4Pool := pool.v4; v4Pool != nil {
 			if m.isRestoreFinishedLocked(IPv4) {
-				v4Pool.releaseExcessCIDRsMultiPool(neededIPs.IPv4Addrs)
+				// releaseExcessCIDRsMultiPool interprets neededIPs as how many
+				// free addresses must remain after a CIDR is dropped.
+				// Therefore we subtract the number of in-use addresses from neededIPs.
+				freeNeeded4 := neededIPs.IPv4Addrs - v4Pool.inUseIPCount()
+				if freeNeeded4 < 0 {
+					freeNeeded4 = 0
+				}
```
Comment (tklauser) [inline pkg/ipam/multipool.go]: Instead of summing up `inUseIPCount` in `(*multiPoolManager).computeNeededIPsPerPoolLocked` (which is called to populate `neededIPS` only to subtract it here again, how about just excluding them from the calculation in `(*multiPoolManager).computeNeededIPsPerPoolLocked`?

**Review by tklauser — COMMENTED** [review]
One comment inline re. avoiding adding up `inUseIPs` only to subtract them again later. Would also be nice to have some unit test coverage for this particular corner case. Have you looked into extending or adding unit tests to cover this case?

File: pkg/ipam/multipool.go (hunk 1 above)
Comment (alimehrabikoshki) [reply pkg/ipam/multipool.go]: If my understanding is correct, excluding inUse globally would break the calculation used for pool replenishment. 

The logic depends on this formula:
```
neededIPs = in-use  + pending  + pre-allocate     (rounded up to preAllocate)
```

pkg/ipam/allocator/multipool/pool_allocator.go L327
```
	for _, reqPool := range cn.Spec.IPAM.Pools.Requested {
		allocatedCIDRs := p.nodes[cn.Name][reqPool.Pool]

		if option.Config.EnableIPv4 {
			neededIPv4Addrs := big.NewInt(int64(reqPool.Needed.IPv4Addrs))
			toAllocate := neededIPv4Addrs.Sub(neededIPv4Addrs, allocatedCIDRs.v4.availableAddrs())

			allocErr := p.allocateCIDRs(cn.Name, reqPool.Pool, ipam.IPv4, toAllocate)
```

allocatedCIDRs.v4.availableAddrs() returns the total capacity of the CIDRs already assigned to the node (used + free). The allocator subtracts that capacity from Needed and the remainder is what the operator tries to hand out in additional CIDRs.

If we no longer count in-use:
```
Needed = pending + pre-allocate  
toAllocate = (pending + preAlloc) – capacity
```
Unless `pending + pre-alloc` exceeds `in-use + free`, toAllocate becomes <= 0, so the operator allocates nothing even though the node still needs capacity.


>Would also be nice to have some unit test coverage for this particular corner case. Have you looked into extending or adding unit tests to cover this case?

I had a look at the existing test and saw that it only uses /24's  and /27's (and/96, /123 for IPv6). I saw the test function was quite large so I shied away from adding one for this case, but I can take a look at extending it.

Comment (alimehrabikoshki) [reply pkg/ipam/multipool.go]: @tklauser I've added a couple tests for `preAlloc ==0` and `preAlloc == 1` - let me know if that looks good to you.

Comment (tklauser) [reply pkg/ipam/multipool.go]: @alimehrabikoshki Thanks for the explanation and adding the test. The proposed solution LGTM.

Comment (tklauser) [inline pkg/ipam/multipool.go]: nit: this could be shortened to

```suggestion
				freeNeeded4 := max(neededIPs.IPv4Addrs - v4Pool.inUseIPCount(), 0)
```

File: pkg/ipam/multipool.go [hunk 2]
```diff
@@ -561,7 +568,11 @@ func (m *multiPoolManager) updateLocalNode(ctx context.Context) error {
 		}
 		if v6Pool := pool.v6; v6Pool != nil {
 			if m.isRestoreFinishedLocked(IPv6) {
-				v6Pool.releaseExcessCIDRsMultiPool(neededIPs.IPv6Addrs)
+				freeNeeded6 := neededIPs.IPv6Addrs - v6Pool.inUseIPCount()
+				if freeNeeded6 < 0 {
+					freeNeeded6 = 0
+				}
```
Comment (tklauser) [inline pkg/ipam/multipool.go]: same here:

```suggestion
				freeNeeded6 := max(neededIPs.IPv6Addrs - v6Pool.inUseIPCount(), 0)
```

**Review by tklauser — APPROVED** [review]
Thanks for the fix andfor  adding the tests! Two nits inline, but probably only worth addressing case there are other things to change.

File: pkg/ipam/multipool.go (hunk 1 above)
Comment (alimehrabikoshki) [reply pkg/ipam/multipool.go]: Yeah that's a lot cleaner thanks. I made the change and force pushed before I saw I could apply the suggestion from here, so it cancelled your auto-merge - sorry about that.

Comment (tklauser) [reply pkg/ipam/multipool.go]: No worries. I've re-triggered all tests and re-enabled auto-merge. Thanks again for your contribution!

Comment (tklauser) [reply pkg/ipam/multipool.go]: @alimehrabikoshki looks like another change is needed to please the Go formatting linter: https://github.com/cilium/cilium/actions/runs/16370115570/job/46258581876?pr=40393 (running `gofmt -w pkg/ipam/multipool.go` should do).

Comment (alimehrabikoshki) [reply pkg/ipam/multipool.go]: ahhh sorry, I keep forgetting to run it. Really need to integrate fmt with my git workflow. Should be done now. Thanks for the review 👍
